File: examples/basic/basic.py
#!/usr/bin/env python3
"""A tour of the Python API.

Run it from the repository root, after building the shared library:

    make build
    make example-py       # or: PYTHONPATH=python python3 examples/basic/basic.py

It writes ./db/example-py.nsq (and ./db/example-py.nsq.trace) and leaves both
behind when it exits, so you can inspect them afterwards:

    cat db/example-py.nsq.trace
    ./bin/nsq stat db/example-py.nsq
"""

import os
import sys

# Make `import nosqlite` work straight from a source checkout, without
# installing anything. This file lives in examples/basic/, so the package
# directory is two levels up.
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "python"))

from nosqlite import Database, NoSQLiteError  # noqa: E402

DIR = "./db"
PATH = DIR + "/example-py.nsq"


def main() -> None:
    # exist_ok=True makes this a no-op when the directory is already there,
    # instead of raising FileExistsError.
    os.makedirs(DIR, exist_ok=True)

    # Start from scratch each run, so the numbers printed below are predictable
    # -- otherwise every run would append another five people to the same
    # collection. Nothing is deleted at the end: both files stay on disk.
    for leftover in (PATH, PATH + ".trace"):
        if os.path.exists(leftover):
            os.remove(leftover)

    # The context manager closes the database on the way out, including if an
    # exception is raised.
    with Database(PATH, trace="all") as db:
        users = db["users"]

        # --- inserting ----------------------------------------------------

        ada_id = users.insert(
            {
                "name": "Ada",
                "age": 36,
                "tags": ["math", "code"],
                "address": {"city": "London"},
            }
        )
        print("inserted Ada with _id:", ada_id)

        users.insert_many(
            [
                {"name": "Grace", "age": 45, "address": {"city": "New York"}},
                {"name": "Alan", "age": 41, "tags": ["math"]},
                {"name": "Edsger", "age": 41, "address": {"city": "Austin"}},
                {"name": "Barbara", "age": 63, "tags": ["biology", "code"]},
            ]
        )
        print("collection now holds:", len(users), "documents")

        # --- querying -----------------------------------------------------

        print("\n-- everyone 40 or older, oldest first --")
        for u in users.find({"age": {"$gte": 40}}, sort=[("age", -1)], limit=10):
            # Numbers come back as Python ints here: json.loads undoes the
            # float64 round trip that happens inside Go.
            print(f"  {u['name']:<8} {u['age']}")

        print("\n-- anyone tagged 'math' (arrays match element-wise) --")
        for u in users.find({"tags": "math"}):
            print(" ", u["name"])

        print("\n-- dotted paths reach into nested objects --")
        # find_one returns None when nothing matches, so it has to be checked
        # before subscripting -- a type checker insists, and it is right to.
        austin = users.find_one({"address.city": "Austin"})
        print(" ", austin["name"] if austin else "nobody", "lives in Austin")

        print("\n-- $or, and counting --")
        n = users.count({"$or": [{"age": {"$lt": 40}}, {"age": {"$gt": 60}}]})
        print(f"  {n} people are under 40 or over 60")

        print("\n-- find_one on no match returns None --")
        print(" ", users.find_one({"name": "Nobody"}))

        # --- streaming ------------------------------------------------------

        # iter_find pages under the hood, so Python-side memory stays flat no
        # matter how many documents match.
        print("\n-- streaming with iter_find --")
        total = sum(u["age"] for u in users.iter_find(batch=2))
        print(f"  average age: {total / len(users):.1f}")

        # --- errors ---------------------------------------------------------

        print("\n-- errors come back as exceptions --")
        try:
            users.find({"age": {"$gtee": 30}})  # typo in the operator
        except NoSQLiteError as exc:
            print(" ", exc)

        print("\ncollections in this database:", db.collections())

    print("\nDatabase file:", PATH)
    print("Trace of everything above is in", PATH + ".trace")
    print("Both files are left in place — try: cat " + PATH + ".trace")
    print("Try: NOSQLITE_TRACE=verbose python3 examples/basic/basic.py")


if __name__ == "__main__":
    main()

File: examples/basic/basic.ts
/**
 * A tour of the TypeScript API.
 *
 * Run it from the repository root, after building the shared library:
 *
 *     make build
 *     make example-ts       # or: node examples/basic/basic.ts
 *
 * There is no compile step: Node >= 22.18 strips the type annotations and runs
 * this file as-is. The only dependency is `koffi` (the FFI), installed by
 * `make example-ts` into typescript/node_modules.
 *
 * The two-line package.json next to this file exists only to say
 * `"type": "module"`. Node decides whether a file is an ES module (`import`)
 * or a CommonJS one (`require`) from the nearest package.json, and the repo
 * root has none — it is a Go project. Without it, tooling reads this file as
 * CommonJS and objects to the `import` lines below. Go and Python ignore it.
 *
 * It writes ./db/example-ts.nsq (and ./db/example-ts.nsq.trace) and leaves both
 * behind when it exits, so you can inspect them afterwards:
 *
 *     cat db/example-ts.nsq.trace
 *     ./bin/nsq stat db/example-ts.nsq
 */

import { mkdirSync, rmSync } from "node:fs";

// Imported by relative path so the example runs straight out of a checkout
// with nothing installed. The `.ts` extension is required: Node resolves the
// file literally, it does not guess extensions the way a bundler would.
import { Database, NoSQLiteError, type Document } from "../../typescript/nosqlite/index.ts";

const DIR = "./db";
const PATH = `${DIR}/example-ts.nsq`;

function main(): void {
  // recursive: true creates missing parents and — unlike the plain call —
  // succeeds when the directory is already there, so it is safe every run.
  mkdirSync(DIR, { recursive: true });

  // Start from scratch each run, so the numbers printed below are predictable
  // — otherwise every run would append another five people to the same
  // collection. Nothing is deleted at the end: both files stay on disk.
  for (const leftover of [PATH, `${PATH}.trace`]) {
    rmSync(leftover, { force: true }); // force: no error if it is not there
  }

  const db = new Database(PATH, { trace: "all" });
  // try/finally is this language's `defer db.Close()`: the database is closed
  // however we leave the block, exception or not.
  try {
    const users = db.collection("users");

    // --- inserting --------------------------------------------------------

    const adaId = users.insert({
      name: "Ada",
      age: 36,
      tags: ["math", "code"],
      address: { city: "London" },
    });
    console.log("inserted Ada with _id:", adaId);

    users.insertMany([
      { name: "Grace", age: 45, address: { city: "New York" } },
      { name: "Alan", age: 41, tags: ["math"] },
      { name: "Edsger", age: 41, address: { city: "Austin" } },
      { name: "Barbara", age: 63, tags: ["biology", "code"] },
    ]);
    console.log("collection now holds:", users.count(), "documents");

    // --- querying ---------------------------------------------------------

    console.log("\n-- everyone 40 or older, oldest first --");
    for (const u of users.find({ age: { $gte: 40 } }, { sort: [["age", -1]], limit: 10 })) {
      // Documents are plain objects with `unknown` values — the database
      // cannot know your schema. padEnd lines the names up.
      console.log(` ${String(u.name).padEnd(8)} ${u.age}`);
    }

    console.log("\n-- anyone tagged 'math' (arrays match element-wise) --");
    for (const u of users.find({ tags: "math" })) {
      console.log(" ", u.name);
    }

    console.log("\n-- dotted paths reach into nested objects --");
    const austin: Document | null = users.findOne({ "address.city": "Austin" });
    console.log(" ", austin?.name, "lives in Austin");

    console.log("\n-- $or, and counting --");
    const n = users.count({ $or: [{ age: { $lt: 40 } }, { age: { $gt: 60 } }] });
    console.log(` ${n} people are under 40 or over 60`);

    console.log("\n-- findOne on no match returns null --");
    console.log(" ", users.findOne({ name: "Nobody" }));

    // --- streaming ----------------------------------------------------------

    // iterFind pages under the hood, so JS-side memory stays flat no matter
    // how many documents match.
    console.log("\n-- streaming with iterFind --");
    let total = 0;
    for (const u of users.iterFind({}, { batch: 2 })) {
      total += Number(u.age);
    }
    console.log(`  average age: ${(total / users.count()).toFixed(1)}`);

    // --- errors -------------------------------------------------------------

    console.log("\n-- errors come back as exceptions --");
    try {
      users.find({ age: { $gtee: 30 } }); // typo in the operator
    } catch (err) {
      // `catch` gives you `unknown`, so narrow before using it.
      if (!(err instanceof NoSQLiteError)) throw err;
      console.log(" ", err.message);
    }

    console.log("\ncollections in this database:", db.collections());
  } finally {
    db.close();
  }

  console.log("\nDatabase file:", PATH);
  console.log("Trace of everything above is in", `${PATH}.trace`);
  console.log(`Both files are left in place — try: cat ${PATH}.trace`);
  console.log("Try: NOSQLITE_TRACE=verbose node examples/basic/basic.ts");
}

main();
